Java/Fundamentals/OOPs Principals/OOPs_Principals_documentation.md

```diff
@@ -0,0 +1,88 @@
+<!-- Image of the states and behaviour -->
+----------
+## What is an object ?
+- Object are the basic unit of OOP with behaviour and identity. As we have told earlier that class is nothing but a blueprint of the object.
+- Classes only act as a template for a physical entity known as **object** .
+- So classes act as logical description for an object and Object is physical instance of the class .
+- There are three steps to creating an object of a class 
+    - **Declaration** :- A variable declaration with a variable name with an object type.
+    - **Instantiation** :-The '**new**' keyword is used to create object in heap memory (We will talk more on heap memory in memory allocation section).
+    - **Initialization** :- After the '**new**' keyword we call the constructor .This initializes the new object.
+- **Syntax** of defining object : 
+<!-- Syntax of creating object -->
+![object](https://github.com/amitx007/winter-of-contributing/blob/Java/Java/Fundamentals/Assets/objectsyntax.jpg)
+
+- Once we have created an object, then to access these class members we will use a special operator that is **Dot operator**.
+- **Syntax** of using dot operator:
+ <!-- Dot operator  -->
+![dot operator](https://github.com/amitx007/winter-of-contributing/blob/Java/Java/Fundamentals/Assets/dotoperator.jpg)
+----------
+##Difference between Classes and Objects in Java
+|Class                                                         |Object                                                 |
+|--------------------------------------------------------------|-------------------------------------------------------|
+|Classes is a logical description of an object                 |Object is a physical instance of the class             |
+|A class is datatype.                                          |Objects are stored in a reference variable of type class|
+|Class is created only once                                    |But we can create as many objects according to our need |
+|Class is created using keyword **class**                     |But object can be created in many ways and one of them is using **new**          keyword| 
+|Classes do not occupy memory location                         |But object occupy memory location  |
+|Classes can not be manipulated as its not available in memory |But we can manipulate objects|
+
+-----------
+## Memory Allocation
+- Memory Allocation is a process by which computer programs are assigned with some memory space .
+- As we know that object's address are stored in the reference variable of type class. Then to which place this address refer to ? Its the address of a space taken by the object in heap memory (which we will talk in this section) and that address is stored in reference variable. Can we print this address? **NO**, we can't print the address of the object but we can print the identity hashcode of that object by writing 
+```Java
+    object.toString();
```

Comment:
I think it is `object.hashCode()`. See [Object Class JavaDocs](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode())

Reply:
Done